scripts/test/build_headers.sh

```diff
@@ -3,42 +3,129 @@
 set -e
 set -o pipefail
 
-mkdir -p test_headers
-rm -f test_headers/Makefile
+# This script depends on the warning_flags file built by configure.
 
-cat >> test_headers/Makefile <<EOF
+EXIT_STATUS=0
 
-all: objects
+# The first argument is $abs_top_srcdir, defaulting to the current directory.
+SOURCE_DIR="${1:-.}"
+# The second argument is $abs_top_builddir, defaulting to the current
+# directory.
+BUILD_DIR="${2:-.}"
+# The third argument is $MAKE, defaulting to "make"
+MAKE_CMD="${3:-make}"
+# Paths in the Makefile are relative to $TEST_DIR
+TEST_DIR="${BUILD_DIR}"/test_headers
+# Headers that are allowed to fail
+EXCEPTIONS_FILE="$SOURCE_DIR"/scripts/test/build-headers-exceptions.txt
 
-%.o: %.c
-	gcc -c -Wall -Werror @../warning_flags -I .. -I ../src -I ../src/ext -o \$@ \$<
+# Paths in this script must be absolute
+RELATIVE_PATH_FAIL_DIR=$(mktemp -d -t build_headers_path_fail_XXXXXX)
+cd "$RELATIVE_PATH_FAIL_DIR"
+chmod a-rwx "$RELATIVE_PATH_FAIL_DIR"
+
+mkdir -p "$TEST_DIR"
+
+cat > "${TEST_DIR}"/prefix.h <<EOF
+/*
+ * List of tor prefix headers
+ * We include this header before testing that other headers can compile by
+ * themselves. Other headers can expect these headers to be included in
+ * every file, before the header itself is included.
+ */
+#include "orconfig.h"
+
+/* TODO: remove these dependencies */
+#include "lib/cc/torint.h"
+#include "lib/testsupport/testsupport.h"
+EOF
+
+# Includes are relative to $abs_top_srcdir{,/src,/src/ext}
+# We also add $BUILD_DIR to the include path, for test_headers/prefix.h
+# (We don't add $TEST_DIR to the include path, to avoid including any other
+# generated files.)
+MAKEFILE_PATH="${TEST_DIR}"/Makefile
+cat > "$MAKEFILE_PATH" <<EOF
+
+all:
+	rm -f header_*.txt
+	\$(MAKE) objects
+
+update-exceptions:
+	\$(MAKE) all
+	sort header_fail.txt > "$EXCEPTIONS_FILE"
+
+.c.o:
+	echo "\$<" >> header_all.txt
+	if gcc -c -Wall -Werror @"${BUILD_DIR}"/warning_flags -Wno-unused-function \
+		-I "${SOURCE_DIR}" -I "${SOURCE_DIR}"/src -I "${SOURCE_DIR}"/src/ext \
+		-I "${SOURCE_DIR}"/src/ext/trunnel -I "${BUILD_DIR}" \
+		-o "\$@" "\$<"; then \
+		echo "\$<" >> header_success.txt; \
+	else \
+		echo "\$<" >> header_fail.txt; \
+	fi
 EOF
 
 add_obj()
 {
     NAME="$1"
     HDR="$2"
-    cat >> test_headers/Makefile <<EOF
+    cat >> "$MAKEFILE_PATH" <<EOF
 objects: ${NAME}.o
-${NAME}.o: ${NAME}.c ${HDR}
+${NAME}.o: ${NAME}.c ${HDR} prefix.h
 EOF
 }
 
+# We deliberately skip win32: it only contains a Windows orconfig.h
+for hdr in $(cd "${SOURCE_DIR}"/src && find lib core feature app test tools -name '*.h'); do
+    hdr_path="${SOURCE_DIR}"/src/"${hdr}"
+    name=$(basename "$hdr" .h)
+    tc_name="${name}"_h
+    tc_path="${TEST_DIR}"/"$tc_name".c
 
-for hdr in $(cd ./src && find lib core feature app -name '*.h'); do
-  name=$(basename "$hdr" .h)
+    grep '^ *# *if\(def *\| .*defined(\)[A-Z_]*\(PRIVATE\|INTERNAL\|EXPOSE\)[A-Z_]*' "$hdr_path" | \
```

Comment:
I thought that we were going to standardize on only using PRIVATE.  Do we have a ticket for that?

Reply:
I'm happy to standardise on [A-Z_]*_PRIVATE. I don't know if we have a ticket for that. (Maybe we should also remove the unused PRIVATE defined at the same time.)

When that ticket merges, we can simplify this pattern, and the exclude list.

Reply:
I opened trac 32818 for this change.